### PR TITLE
Add SWO support for MIMXRT1170-EVK

### PR DESCRIPTION
**Pull request by groleo**
This p-r is adding support for NXP's i.MXRT1170-EVK board
and enables SWO for it.

### CODE REVIEW

**Comment by groleo** [comment]
cc @flit @maclobdell @marcuschangarm @0xc0170

**Comment by 0xc0170** [comment]
I'll review this PR today or tomorrow

**Comment by 0xc0170** [comment]
There's build error:

```
04:41:15 Building project daplink (lpc4322_mimxrt1170_evk_qspi_if, ARM)
04:41:15 Scan: daplink
04:41:15 Compile [  1.3%]: mimxrt1170_evk.c
04:41:15 Compile [  2.6%]: HardFault_Handler.c
04:41:15 Compile [  3.8%]: circ_buf.c
04:41:15 Compile [  5.1%]: DAP_queue.c
04:41:15 Compile [  6.4%]: JTAG_DP.c
04:41:15 Compile [  7.7%]: SWO.c
04:41:15 [Error] SWO.c@31,27:  #5: cannot open source input file "Driver_USART.h": No such file or directory
04:41:16 [ERROR] ".\source\daplink\cmsis-dap\SWO.c", line 31: Error:  #5: cannot open source input file "Driver_USART.h": No such file or directory
04:41:16   #include "Driver_USART.h"
```

**Comment by 0xc0170** [comment]
Can we reuse the files from KL26 or move them as suggested in the issue?

**Comment by groleo** [comment]
> There's build error:
fixed. I forgot to add the RTE_Driver files.

**Comment by groleo** [comment]
> 
> 
> Can we reuse the files from KL26 or move them as suggested in the issue?

Sure. Let me know where you want to move the Driver_Common.h and Driver_USART.h headers from K26F HIC folder.

I don't know if the RTE_Driver files I've added should also be common.

**Comment by askyuan** [comment]
study

File: source/family/nxp/mimxrt1170_spi_flash/DAP_config.h [hunk 1]
```diff
@@ -0,0 +1,548 @@
...
+
+/// Maximum Package Buffers for Command and Response data.
+/// This configuration settings is used to optimized the communication performance with the
+/// debugger and depends on the USB peripheral. For devices with limited RAM or USB buffer the
+/// setting can be reduced (valid range is 1 .. 255). Change setting to 4 for High-Speed USB.
+#define DAP_PACKET_COUNT      1              ///< Buffers: 64 = Full-Speed, 4 = High-Speed.
+
+/// Indicate that UART Serial Wire Output (SWO) trace is available.
+/// This information is returned by the command \ref DAP_Info as part of <b>Capabilities</b>.
+#define SWO_UART                1               ///< SWO UART:  1 = available, 0 = not available
```
Comment (flit) [inline source/family/nxp/mimxrt1170_spi_flash/DAP_config.h:96]: The `DAP_config.h` file should be shared between all uses of the LPC4322, not specific to a target. If you want to customize per target, you can wrap this line with `#if !defined(SWO_UART)`, then define that macro in the target project yaml.

File: records/board/mimxrt1170_evk_qspi.yaml [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+        - SOFT_RESET=VECTRESET
+    includes:
+        - source/family/nxp/mimxrt1170_spi_flash
+        - source/hic_hal/nxp/lpc4322/RTE_Driver
+        - source/hic_hal/nxp/lpc4322/RTE_Driver/Config
+    sources:
+        lpc4322:
+            - source/hic_hal/nxp/lpc4322/RTE_Driver/USART_LPC43xx.c
+            - source/hic_hal/nxp/lpc4322/RTE_Driver/GPDMA_LPC43xx.c
+            - source/hic_hal/nxp/lpc4322/RTE_Driver/SCU_LPC43xx.c
```
Comment (flit) [inline records/board/mimxrt1170_evk_qspi.yaml:12]: Please add these sources and includes to the `lpc4322.yaml` so they are common.

**Comment by flit** [comment]
Hi @groleo, would you mind removing the RTE_Driver sources that aren't used? To answer your question, these files are specific to the LPC4322, so should stay under that folder.

And yes, please move the `Driver_Common.h` and `Driver_USART.h` headers. For now, please place them directly under `source/hic_hal/`. Or do you think `source/cmsis-driver/` would be better?

**Comment by flit** [comment]
cc @mmahadevan108

**Comment by groleo** [comment]
Hi,

Got it. I'll move the lpc specific files.

If you ask me, it makes sense to have the "Driver_*" under
source/cmsis-driver.

The files are coming from a sub-directory with the same name, part of a
Kyle lpc-link2 sample.

I'll push  a new p-r tomorow.

**Comment by flit** [comment]
You don't need to push a new PR (please don't). Just force push (assuming you even need to) from the same branch, and it will update this PR.

Let's put the CMSIS-Driver headers under `source/cmsis-driver` as you prefer.

**Comment by groleo** [comment]
The ci-morph "Details" link is pointing to an internal IP:

     http://10.118.80.30:8080/job/daplink_build_mbedcli_pr/257/

Is there another way I can see what went wrong ?
thx

**Comment by groleo** [comment]
v2: CMSIS-Driver headers under source/cmsis-driver as you prefer.
v3: reorder commits (first reorder the headers, then add platform specific changes)

**Comment by groleo** [comment]
> 
> 
> /morph test

Is there a way to see the CI failures ?

**Comment by 0xc0170** [comment]
It's internal server, I'll find out now

**Comment by 0xc0170** [comment]
```
10:56:38 Compile [ 50.0%]: SCU_LPC43xx.c
10:56:38 [Error] SCU_LPC43xx.c@33,26:  #5: cannot open source input file "SCU_LPC43xx.h": No such file or directory
10:56:39 [ERROR] ".\source\hic_hal\nxp\lpc4322\RTE_Driver\SCU_LPC43xx.c", line 33: Error:  #5: cannot open source input file "SCU_LPC43xx.h": No such file or directory
10:56:39   #include "SCU_LPC43xx.h"
10:56:39                           ^
10:56:39 .\source\hic_hal\nxp\lpc4322\RTE_Driver\SCU_LPC43xx.c: 0 warnings, 1 error
```

This is the building error.

**Comment by groleo** [comment]
dough!
v2: CMSIS-Driver headers under source/cmsis-driver as you prefer.
v3: reorder commits (first reorder the headers, then add platform specific changes)
v4: git add the missing RTE_Driver header files (GPDMA_LPC43xx.h, USART_LPC43xx.h and SCU_LPC43xx.h)

**Comment by groleo** [comment]
Ping?

**Comment by 0xc0170** [comment]
Hope the testing will go through this time.

**Review by 0xc0170 — APPROVED** [review]

**Comment by groleo** [comment]
v2: CMSIS-Driver headers under source/cmsis-driver as you prefer.
v3: reorder commits (first reorder the headers, then add platform specific changes)
v4: git add the missing RTE_Driver header files (GPDMA_LPC43xx.h, USART_LPC43xx.h and SCU_LPC43xx.h)
v5: add source/cmsis-driver include to records/hic_hal/lpc4322.yaml (previously it was in mimxrt1170_evk_qspi.yaml)

**Comment by groleo** [comment]
v6: fix the flash-algo bits (Alex Yang alex.yang@nxp.com)
v5: add source/cmsis-driver include to records/hic_hal/lpc4322.yaml (previously it was in mimxrt1170_evk_qspi.yaml)
v4: git add the missing RTE_Driver header files (GPDMA_LPC43xx.h, USART_LPC43xx.h and SCU_LPC43xx.h)
v3: reorder commits (first reorder the headers, then add platform specific changes)
v2: CMSIS-Driver headers under source/cmsis-driver as you prefer.

**Comment by groleo** [comment]
Hi,

What is the ci-morph ?
thx

**Comment by groleo** [comment]
v7:
- increase the system clock from 96MHz to 120MHz.
- increase DAP_PACKET_SIZE and DAP_PACKET_COUNT to take advantage of the high-speed mode;
   I see an increase of IAR's transfer rate from 150 kB/sec to 330 kB/sec.
- implement gpio_set_board_power and LED_CONNECTED_OUT
- use the gpio X_* helpers instead of going through LPC_GPIO_PORT
v6: fix the flash-algo bits (Alex Yang alex.yang@nxp.com)
v5: add source/cmsis-driver include to records/hic_hal/lpc4322.yaml (previously it was in mimxrt1170_evk_qspi.yaml)
v4: git add the missing RTE_Driver header files (GPDMA_LPC43xx.h, USART_LPC43xx.h and SCU_LPC43xx.h)
v3: reorder commits (first reorder the headers, then add platform specific changes)
v2: CMSIS-Driver headers under source/cmsis-driver as you prefer.

File: records/board/mimxrt1170_evk_qspi.yaml [hunk 3]
```diff
@@ -0,0 +1,13 @@
+common:
+    macros:
+        - SOFT_RESET=VECTRESET
+        - SWO_UART=1
+        - SWO_USART_PORT=1
```
Comment (flit) [inline records/board/mimxrt1170_evk_qspi.yaml:5]: I don't know that it matters, but I have been putting the `SWO_UART_*` macro definitions in `DAP_config.h`.

File: source/board/mimxrt1170_evk.c [hunk 4]
```diff
@@ -0,0 +1,33 @@
+/**
+ * @file    mimxrt1170_evk.c
+ * @brief   board ID for the NXP MIMXRT1170-EVK board.
+ *
+ * DAPLink Interface Firmware
+ * Copyright (c) 2009-2020, ARM Limited, All Rights Reserved
```
Comment (flit) [inline source/board/mimxrt1170_evk.c:6]: Please change the copyright to your own (or NXP's, as appropriate) for 2020-2021.

File: source/daplink/drag-n-drop/flash_manager.c [hunk 5]
```diff
@@ -48,7 +48,7 @@ __attribute__((aligned(4)))
 static uint8_t buf[1024];
 static bool buf_empty;
 static bool current_sector_valid;
-static bool page_erase_enabled = false;
+static bool page_erase_enabled = true;
```
Comment (flit) [inline source/daplink/drag-n-drop/flash_manager.c:51]: I'm slightly concerned about making this change globally, only because most of the flash algos have not been tested in page erase mode. You can use set the `kEnablePageErase` flag in the `board_info_t` struct to set it per board. It's _probably_ safe to change it globally, but that would better be done in a separate PR.

File: source/family/freescale/target_reset_mimxrt.c [hunk 6]
```diff
@@ -47,14 +47,21 @@ static void prerun_target_config(void)
...
     }
-    else
+    else if(buf[0] == 0xFF && buf[1] == 0xFF && buf[2] == 0xFF && buf[3] == 0xFF)
     {
-        return 0;
+        target_device.flash_regions[0].start = 0x30000000;
+        return 1;
     }
+
+    return 0;
```
Comment (flit) [inline source/family/freescale/target_reset_mimxrt.c:64]: Nice hack! Would you mind adding some comments describing what this code does so we don't forget in the future? 😄

**Review by flit — CHANGES_REQUESTED** [review]
The only changes requested are updating the copyrights, page erase mode default, and possibly enabling SWO on all LPC4322 based boards.

**Comment by flit** [comment]
@groleo My apologies this PR has been outstanding for so long! Honestly I haven't been paying much attention since I've been primarily focused on pyocd.

File: source/daplink/drag-n-drop/flash_manager.c (hunk 5 above)
Comment (groleo) [reply source/daplink/drag-n-drop/flash_manager.c]: added kEnablePageErase to source/board/mimxrt1170_evk.c.

File: source/board/mimxrt1170_evk.c (hunk 4 above)
Comment (groleo) [reply source/board/mimxrt1170_evk.c]: fixed as suggested.

File: source/family/freescale/target_reset_mimxrt.c (hunk 6 above)
Comment (groleo) [reply source/family/freescale/target_reset_mimxrt.c:64]: From what I understand, if buf is a FlexSPI Config block (the "FCFB"/0x42464346 test), it moves the flash start address;
but I have no idea what the test against 0xFFFFFFFF part does; I see some other platforms have a "cookieList"

**Comment by mbrossard** [comment]
Superseded by #906.
